_posts/2020-11-12-symbolic-execution-with-ds-test.md

```diff
@@ -0,0 +1,421 @@
+introduced symbolic execution features into its unit testing framework.
+
+This lets users of the [`dapp`](https://github.com/dapphub/dapptools/tree/master/src/dapp) smart
+contract development framework prove properties about their contracts using the same syntax and
+language features that they use to write the contracts themselves.
+
+In this tutorial we will show how to use these new features to prove properties of your smart
+contracts.
+
+---
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**
+
+- [Installation](#installation)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Installation
+
+If you want to follow along with the examples in this blog post, you need to install the
+[`nix`](https://nixos.org/) package manager, and then install the
+[`dapptools`](https://github.com/dapphub/dapptools) toolkit:
+
+```sh
+# user must be in sudoers
+curl -L https://nixos.org/nix/install | sh
+
+# Run this or login again to use Nix
+. "$HOME/.nix-profile/etc/profile.d/nix.sh"
+
+# install dapptools
+curl https://dapp.tools/install | sh
+```
+
+## What Is Symbolic Execution?
+
+Symbolic execution is a program analysis technique that executes programs on symbolic inputs,
+representing many possible concrete inputs, and represents the program behavior using mathematical
+constraints over the symbolic inputs.
+
+To make this a little more concrete, consider the following contract:
+
+```solidity
+contract Add {
+    function add(uint x, uint y) external pure returns (uint z) {
+        require((z = x + y) >= x, "overflow!");
+    }
+}
+```
+
+We can represent the behaviour of the `add` method as a tree of possible executions:
+
+```
+├ 0     msg.value > 0
+│       Revert
+│
+└ 1     msg.value == 0
+  │
+  ├ 0   x + y < x
+  │     Revert("overflow!")
+  │
+  └ 1   x + y >= x
+        Return: x + y
+```
+
+Each leaf on the tree represents a possible execution path, and has some logical conditions
+attached:
+
+- `0`:  `msg.value > 0`: revert (`add` is not payable)
+- `10`: `msg.value == 0 && x + y < x`: revert (overflow)
+- `11`: `msg.value == 0 && x + y >= x`: return `x + y`
+
+We can therefore use symbolic execution to exhaustively explore all possible execution paths for a
+given program. This means we can write tests that assert that properties hold for *all possible
+inputs* to a given function.
+
+## Using `ds-test`
+
+[`ds-test`](https://github.com/dapphub/ds-test/blob/master/src/test.sol) is a solidity unit testing
+libary with tight integration into `hevm`. `hevm` will execute as a test any method that meets the
+following criteria:
+
+1. The method is on a contract that inherits from `DSTest`
+1. The method is `public` or `external`
+1. The method name starts with `test` or `prove`
+
+If a public method named `setUp()` is present it will be executed before each test is run.
```

Comment:
Note: 
Currently, `setUp` is actually required right now. It should be made optional tho